src/shader/shaderLightSampling.js

```diff
@@ -127,66 +161,67 @@ LightSampleRec randomAreaLightSample( Light light, vec3 rayOrigin ) {
 
 }
 
-LightSampleRec randomSpotLightSample( SpotLight spotLight, sampler2DArray iesProfiles, vec3 rayOrigin ) {
-
-	LightSampleRec lightSampleRec;
-	lightSampleRec.hit = true;
+LightSampleRec randomSpotLightSample( Light light, sampler2DArray iesProfiles, vec3 rayOrigin ) {
 
-	float r = spotLight.radius * sqrt( rand() );
+	float radius = light.radius * sqrt( rand() );
 	float theta = rand() * 2.0 * PI;
-	float x = r * cos( theta );
-	float y = r * sin( theta );
+	float x = radius * cos( theta );
+	float y = radius * sin( theta );
 
-	vec3 u = spotLight.u;
-	vec3 v = spotLight.v;
-	vec3 lightNormal = normalize( cross( u, v ) );
+	vec3 u = light.u;
+	vec3 v = light.v;
+	vec3 normal = normalize( cross( u, v ) );
 
-	float angle = acos( spotLight.coneCos );
+	float angle = acos( light.coneCos );
 	float angleTan = tan( angle );
-	float startDistance = spotLight.radius / max( angleTan, EPSILON );
+	float startDistance = light.radius / max( angleTan, EPSILON );
 
-	vec3 randomPos = spotLight.position - lightNormal * startDistance + u * x + v * y;
+	vec3 randomPos = light.position - normal * startDistance + u * x + v * y;
 	vec3 toLight = randomPos - rayOrigin;
 	float lightDistSq = dot( toLight, toLight );
-	lightSampleRec.dist = sqrt( lightDistSq );
-
-	vec3 direction = toLight / max( lightSampleRec.dist, EPSILON );
-	lightSampleRec.direction = direction;
+	float dist = sqrt( lightDistSq );
 
-	float cosTheta = dot( direction, lightNormal );
+	vec3 direction = toLight / max( dist, EPSILON );
+	float cosTheta = dot( direction, normal );
 
-	float spotAttenuation = spotLight.iesProfile != -1 ?
-		  getPhotometricAttenuation( iesProfiles, spotLight.iesProfile, direction, lightNormal, u, v )
-		: getSpotAttenuation( spotLight.coneCos, spotLight.penumbraCos, cosTheta );
+	float spotAttenuation = light.iesProfile != - 1 ?
+		  getPhotometricAttenuation( iesProfiles, light.iesProfile, direction, normal, u, v )
+		: getSpotAttenuation( light.coneCos, light.penumbraCos, cosTheta );
 
-	float distanceAttenuation = getDistanceAttenuation( lightSampleRec.dist, spotLight.distance, spotLight.decay );
-
-	lightSampleRec.emission = spotLight.color * spotLight.intensity * max( distanceAttenuation * spotAttenuation, EPSILON );
+	float distanceAttenuation = getDistanceAttenuation( dist, light.distance, light.decay );
+	float cosTheta = dot( direction, normal );
+	LightSampleRec lightSampleRec;
+	lightSampleRec.hit = true;
+	lightSampleRec.dist = dist;
+	lightSampleRec.direction = direction;
+	lightSampleRec.emission = light.color * light.intensity * distanceAttenuation * spotAttenuation;
 
+	// TODO: this makes the result consistent between MIS and non MIS paths but at radius 0 the pdf is infinite
+	// and the intensity of the light is not correct
 	lightSampleRec.pdf = 1.0;
+	// lightSampleRec.pdf = lightDistSq / ( light.area * cosTheta );
```

Comment:
@richardassar This is where I've adjusted the PDF for the random light sampling. Of course as the radius becomes zero the PDF value becomes Infinity which causes issues. I'm not sure if you have any thoughts here?

Reply:
At one point I had coupled the circular area light with the spot light, in the submitted PR comments (I think) I mentioned that these were decoupled as I thought circular area lights might be useful on their own and that users of spot lights may or may not want to represent the (approximate) diffuse illumination of the lamp.

I think this PDF was fine as it was for a circular (and rectangular) area light. I think simply clamping the denominator to EPS is sufficient here.

Reply:
> I suppose once answer would be to divide some of the PDF equation into the emission?

This seems to work, shifting `( light.area * cosTheta )` over to the emission whilst leaving only `lightDistSq` as the PDF works. I can set a zero radius without problems.

Btw, there was a duplicate of `float cosTheta = dot( direction, normal );` that I just commented out locally as you'd see it in the logs straight away.

Reply:
> At one point I had coupled the circular area light with the spot light, in the submitted PR comments (I think) I mentioned that these were decoupled as I thought circular area lights might be useful on their own and that users of spot lights may or may not want to represent the (approximate) diffuse illumination of the lamp.

Yeah I do recall that - I like the separation we have now since the spotlight coverage is different than that circular area light. And it's nice that soft shadows can be supported by increasing the radius. I'm just trying to grok how this should work with the PDF, MIS, and basic forward path tracing to support the specular reflections.

In the mean time I'd like to keep this simplified code path and merged lighting texture, though. I'm thinking I'll disable spot light support for the non-MIS path for now and make an issue for understanding the details of this in the future. These links suggested on Twitter should be helpful for understanding how to handle this. What do you think of that solution for now?

- https://www.pbr-book.org/3ed-2018/Light_Sources/Point_Lights#Spotlights
- https://www.pbr-book.org/3ed-2018/Light_Transport_I_Surface_Reflection/Sampling_Light_Sources#LightswithSingularities

> Btw, there was a duplicate of float cosTheta = dot( direction, normal ); that I just commented out locally as you'd see it in the logs straight away.

Thanks I noticed that, too. Forgot to commit the fix 😁

Reply:
During development I had encountered https://www.pbr-book.org/3ed-2018/Light_Sources/Point_Lights#fragment-SpotLightMethodDefinitions-1 and some code examples in other projects which echoed this. 

For non-IES lights some examples take the approach of setting the PDF to 1.0 inside and 0.0 outside of the spot light cone, I don't like this as it seems luminosity or intensity should affect importance sampling (as [DassaultSystemes-Technology/dspbr-pt](https://github.com/DassaultSystemes-Technology/dspbr-pt does for the diffuse and specular components, convergence is much faster).

I too had this issue with trying to match everything up and I spent quite a bit of time on it over multiple attempts before taking a different approach. Seems reasonable to drop non-MIS for now in favour of code simplification, then later a fresh set of eyes can close the gap.

Reply:
> I don't like this as it seems luminosity or intensity should affect importance sampling

Yeah I agree - this could implicitly help with #235, as well. I made #247 and #248 to track these issues. I'll merge this with just multiple importance sampling support for now and we can revisit later. Thanks!